### PR TITLE
[NN module] speed up _load_from_state_dict

### LINKED ISSUE #61398: Execution time of load_state_dict
I'm refactoring a model from older torch version (v0.3.1) to newer version (v1.9.0). After that I realized the training time gets very slow. The problem is in `nn.module.load_state_dict`. 
The change happens in #6639, the old code is understandable but new code becomes fancy (I haven't fully understand how does it work) and there is a recursion call in it. Looks like the recursion call is making my model loading parameters 200x slower. The problem is my model have to load parameter for every training step, that's why it becomes slower. 
Therefore:
1. Is it safe to just temporarily copy and use old load_state_dict in newer version? I'm trying to do that, the training speed returns to normal (actually even faster), their results are exactly the same, and no problem occurs so far.
2. Is it safe to request to refine current code in terms of execution time, or implement a alternative lightweight version of `load_state_dict`? I think there must be people who have same problem as me.
Thanks!

cc @albanD @mruberry @jbschlosser @VitalyFedyunin @ngimel @heitorschueroff

### PR DESCRIPTION
**Pull request by zhengwy888**
Fixes #61398

The original implementation is very slow when the state_dict.keys() is long. This PR only passes relevant keys to the child module. 

existing test passes: `pytest test/test_nn.py -k state_dict`
I couldn't figure out a good way to write a new test for this new behavior. Had a new snippet, but it will be flaky if integrated into the main CI because it's a timing based check. 
But I can verify that the test took 30s to run, after this PR it only takes 0.5s. 

```python
    def test_load_state_dict_large(self):
        # construct a module with 4 levels of module, 10 linear each, leads to 10k items in the dictionary
        import copy
        import time
        base_module = nn.Linear(1,1)
        model = base_module
        for level in range(4):
           model = nn.Sequential(*[copy.deepcopy(model) for _ in range(10)])
        state_dict = model.state_dict()
        self.assertEqual(len(state_dict), 20000)
        st = time.time()
        model.load_state_dict(state_dict, strict=True)
        strict_load_time = time.time() - st
        # it took 0.5 seconds to 
        self.assertLess(strict_load_time, 10)
```

### CODE REVIEW

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -1630,13 +1630,15 @@ def load_state_dict(self, state_dict: Mapping[str, Any],
...
             local_metadata = {} if metadata is None else metadata.get(prefix[:-1], {})
             module._load_from_state_dict(
-                state_dict, prefix, local_metadata, True, missing_keys, unexpected_keys, error_msgs)
+                local_state_dict, prefix, local_metadata, True, missing_keys, unexpected_keys, error_msgs)
             for name, child in module._modules.items():
                 if child is not None:
-                    load(child, prefix + name + '.')
+                    child_prefix = prefix + name + '.'
+                    child_state_dict = {k: v for k, v in local_state_dict.items() if k.startswith(child_prefix)}
+                    load(child, child_state_dict, prefix + name + '.')
```
Comment (albanD) [inline torch/nn/modules/module.py:1641]: can you change the last argument to `child_prefix` to avoid duplication please?

**Review by albanD — COMMENTED** [review]
Thanks for sending a fix for this.
This is technically BC-breaking as the submodules will only see a subset of the state dict when loading. And in particular when `Module._register_load_state_dict_pre_hook()` are used (this is not public yet so it is ok to break it).
FYI @rohan-varma 

Using this private hook API, could you add a test that ensures that each of these hooks only see the relevant part of the state dict and not all of it?

**Comment by zhengwy888** [comment]
> Using this private hook API, could you add a test that ensures that each of these hooks only see the relevant part of the state dict and not all of it?

done.

**Review by albanD — APPROVED** [review]
Thanks!

**Comment by albanD** [comment]
> Separately I would like to ask whether we should propagate this strict argument here? currently the caller to _load_from_state_dict() hard code strict=True.

I think this is a bug yes. But we should think about the implication of that change... We should do it in a separate PR for sure.

**Comment by zhengwy888** [comment]
turns out strict=True was added in https://github.com/pytorch/pytorch/pull/18668, and it seems the point is to allow user to handle missing keys separately after the model has been loaded. I am not sure how many people are relying on this function now, too risky to remove...

**Pull request closed without merging**